### ISSUE TITLE
Add disabled property to Control class

### ISSUE THREAD

**Issue by CsloudX** [godotengine/godot-proposals#2844]
### Describe the project you are working on
A no-game application

### Describe the problem or limitation you are having in your project
I have 2 container with a lots controls, user choose 1 to edit. Now I must code like this:
```
  button1.disabled = !is_edit1
  button2.disabled = !is_edit1
  lineEdit1.disabled = !is_edit1
  lineEdit2.disabled = !is_edit1
  button1a.disabled = is_edit1
  button2a.disabled = is_edit1
  lineEdit1a.disabled = is_edit1
  lineEdit2a.disabled = is_edit1
```
In many project, controls was much more.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
If control have diabled property, disable parent control node, all chidren auto disabled.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
If have this feature, I can code like this:
```
con1.disabled = !is_edit1 #con1 container button1,button2,lineEdit1,lineEdit2 nodes as children
conA.disabled = is_edit1 #conA container button1a,button2a,lineEdit1a,lineEdit2a nodes as children
```

**Comment by SilencedPerson** [comment]
> ### Describe the project you are working on
> A no-game application
> 
> ### Describe the problem or limitation you are having in your project
> I have 2 container with a lots controls, user choose 1 to edit. Now I must code like this:
> 
> ```
>   button1.disabled = !is_edit1
>   button2.disabled = !is_edit1
>   lineEdit1.disabled = !is_edit1
>   lineEdit2.disabled = !is_edit1
>   button1a.disabled = is_edit1
>   button2a.disabled = is_edit1
>   lineEdit1a.disabled = is_edit1
>   lineEdit2a.disabled = is_edit1

But why would you ever need to do something like this? The controls already block other controls from receiving input.

**Comment by CsloudX** [comment]
Sorry，@SilencedPerson I doesn't understand your mean.
I say, if have the feature, my code can write like this:
Before:
```
  button1.disabled = !is_edit1
  button2.disabled = !is_edit1
  lineEdit1.disabled = !is_edit1
  lineEdit2.disabled = !is_edit1
  button1a.disabled = is_edit1
  button2a.disabled = is_edit1
  lineEdit1a.disabled = is_edit1
  lineEdit2a.disabled = is_edit1
```
After:
```
con1.disabled = !is_edit1
conA.disabled = is_edit
```
code can much less than before.

**Comment by SilencedPerson** [comment]
Ok. i don't understand why would you ever need to do this. Why do you disable all these buttons anyway when user selects something? It seems completely pointless. Never used program that would do something like this.
If you need to disable this much controls, use popup instead or do it with signals.

I think your need for disabled variable arises from very specific misuse of the Control Nodes as whole.

**Comment by CsloudX** [comment]
@SilencedPerson Like this?

Before:
```
func _on_NewModeButton_clicked(pressed):
  button1.disabled = !pressed
  button2.disabled = !pressed
  lineEdit1.disabled = !pressed
  lineEdit2.disabled = !pressed
  button1a.disabled = pressed
  button2a.disabled = pressed
  lineEdit1a.disabled = pressed
  lineEdit2a.disabled = pressed
```
After:
```
func _on_NewModeButton_clicked(pressed):
  con1.disabled = !pressed
  conA.disabled = pressed
```

**Comment by LikeLakers2** [comment]
@CsloudX The use case you have sounds like it could be solved by the [TabsContainer](https://docs.godotengine.org/en/stable/classes/class_tabcontainer.html) control.

**Comment by Xrayez** [comment]
By the way, we're moving away from `disabled` properties to `enabled`, even if those kind of properties will be enabled by default.

**Comment by CsloudX** [comment]
@LieLakers2 Thanks, use TabsContainer was a good way.
I just think have this feature used often. like this:
![捕获1](https://user-images.githubusercontent.com/35988767/121341141-f3388180-c952-11eb-8b7f-be1eb74e3e8d.PNG)
![捕获2](https://user-images.githubusercontent.com/35988767/121341176-fcc1e980-c952-11eb-8f7c-97b735d42ba4.PNG)
![捕获3](https://user-images.githubusercontent.com/35988767/121341186-ffbcda00-c952-11eb-91d4-f6a60bcf9a3b.PNG)

I can just disabled the GridContainer. not disabled all conainter's chidren edits. maybe in the future, I will add more controls, if GridContainer have disabled property, i'm not need add code.

**Comment by timothyqiu** [comment]
@CsloudX As a workaround, you can do this with a loop instead of one by one:

```gdscript
for control in $GridContainer.get_children():
    control.disabled = true
```

**Comment by CsloudX** [comment]
@timothyqiu Thanks
It need more, because some controls like Label doesn't have disabled property and mybe it is a subcontaniner
```
func disable_children(con):
    for child in con.get_children():
       if child is Container:
           disable_chilrend(child)
       if child.get_property_list().has("disabled"):
           child.disabled = true
```

**Comment by amonroejj** [comment]
This seems like a must-have to me (and is a blocker in my current hobby project: I would like to display box scores in a sports sim and gray out the labels for yet-to-be-played innings or periods, but Label does not have a `disabled` property(!)). The ability to disable/enable really ought to be at the `Control` class level and inherited by everything derived from it.

Graying out a container full of sub-controls with one call is in line with other GUI systems. It would be very useful for apps where you make a lot of long-running web service or database calls, and you need to communicate to the user that the app is busy.

**Comment by pycbouh** [comment]
@amonroejj 

> but Label does not have a disabled property(!)).

Label is not interactive, a property would have no practical effect on it. If you want to dynamically change color of a label, you can use `add_color_override` method with the corresponding theme property, just like with any control.

> Graying out a container full of sub-controls with one call is in line with other GUI systems. It would be very useful for apps where you make a lot of long-running web service or database calls, and you need to communicate to the user that the app is busy.

For something like that you can just add a blocker overlay on top of the controls that you don't want you user to interact with. You can also use node groups and collect your controls using that, making any bulk operation on them trivial.

**Comment by amonroejj** [comment]
> you can just add a blocker overlay

Of course, but it is unnatural, and just a kludge workaround for absent capabilities.

**Comment by Calinou** [comment]
@amonroejj Node has a `propagate_call()` method you can use to set the `disabled` property on a node and all its children:

`$SomeNode.propagate_call("set_disabled", true)`

**Comment by pycbouh** [comment]
> Of course, but it is unnatural, and just a kludge workaround for absent capabilities.

Node groups is what you're supposed to use in those cases, so the capabilities are not at all absent.

**Comment by amonroejj** [comment]
An example from .NET Winforms: It is completely in line with user expectations that labels should change style when disabled.
![gui-disable](https://user-images.githubusercontent.com/2057161/145138908-32aea47e-3518-4468-b803-41e98e24ce5c.gif)

**Comment by CsloudX** [comment]
I'm look at UE, the `Widget` has a `IsEnabled` property, `Widget` was all Control's(like Button, Label etc) and Container's(like HorizontalBox, Grid Pannel etc) parent class.
I'm also look at Qt, the `QWidget` has a `enabled` property, `QWidget` was all Control's(like QPushButton, QLabel etc) and Container's(like QHBoxLayout, QGridLayout etc) parent class.
I'm also look at C# Winform, the `Control` has a `Enabled` property, `Control` was all Control's(like Button, Label etc) and Container's(like GroupBox, Panel etc) parent class.
All of them has the same behavior and I think them was more make sense than Godot. For me, two advantange I like was:
1. Enable/Disable parent control node can auto Enable/Disable chidren nodes.
2. Every Control's has Enabled property, It make sense, for Godot, Label no disable property, it not make sense.

**Comment by CsloudX** [comment]
> > Enable/Disable parent control node can auto Enable/Disable chidren nodes.
> 
> You achieve this in Godot by disabling processing.

For the Control's Enable/Disable property that we discussed, for input control (like button, spinbox etc), it only disable user input and change theme(for now, lineedit and button was ok), for display control(like label, texturrect etc), it only change theme. 
IMO, NOT should disabling processing. and for some display controls(like label), the only thing i think was change it theme.

AND OTHER, I mean, I wish Godot's control's enable/disable can like other UI system, it more make sense.

**Comment by Delsin-Yu** [comment]
> > ### Describe the project you are working on
> > A no-game application
> > ### Describe the problem or limitation you are having in your project
> > I have 2 container with a lots controls, user choose 1 to edit. Now I must code like this:
> > ```
> >   button1.disabled = !is_edit1
> >   button2.disabled = !is_edit1
> >   lineEdit1.disabled = !is_edit1
> >   lineEdit2.disabled = !is_edit1
> >   button1a.disabled = is_edit1
> >   button2a.disabled = is_edit1
> >   lineEdit1a.disabled = is_edit1
> >   lineEdit2a.disabled = is_edit1
> > ```
> 
> But why would you ever need to do something like this? The controls already block other controls from receiving input.

It is necessary. Games often need to support Gamepad / Keyboard Input Styles, it is crucial to ensure UI navigation does not move to an undesired control.

I put up a little demo to demonstrate a common use case.


https://github.com/godotengine/godot-proposals/assets/71481700/bdf19eea-c771-4aa8-8c12-9cd764f394c5
